src/IECore/MeshAlgo.cpp

```diff
@@ -1,16 +1,238 @@
 #include <assert.h>
 #include <algorithm>
 
+#include "boost/mpl/and.hpp"
 #include "OpenEXR/ImathVec.h"
 
 #include "IECore/MeshAlgo.h"
+#include "IECore/FaceVaryingPromotionOp.h"
+#include "IECore/DespatchTypedData.h"
+#include "IECore/DespatchTypedData.h"
+#include "IECore/TypeTraits.h"
 
 using namespace IECore;
 using namespace Imath;
 
 namespace
 {
 
+template< typename T > struct IsArithmeticVectorTypedData : boost::mpl::and_<
+	IECore::TypeTraits::IsNumericBasedVectorTypedData<T>,
+	boost::mpl::not_< IECore::TypeTraits::IsBox<typename TypeTraits::VectorValueType<T>::type > >,
+	boost::mpl::not_< IECore::TypeTraits::IsQuat<typename TypeTraits::VectorValueType<T>::type > >
+> {};
+
+
+struct  AverageValueFromVector
+{
+	typedef DataPtr ReturnType;
+
+	template<typename From> ReturnType operator()( typename From::ConstPtr data )
+	{
+		const typename From::ValueType &src = data->readable();
+		if ( src.size() )
+		{
+			return new TypedData< typename From::ValueType::value_type >( std::accumulate( src.begin() + 1, src.end(), *src.begin() ) / src.size() );
+		}
+		return NULL;
+	}
+};
+
+
+struct MeshVertexToUniform
+{
+	typedef DataPtr ReturnType;
+
+	MeshVertexToUniform( const MeshPrimitive *mesh )	:	m_mesh( mesh )
+	{
+	}
+
+	template<typename From> ReturnType operator()( const From* data )
+	{
+		typename From::Ptr result = static_cast< From* >( Object::create( data->typeId() ).get() );
+		typename From::ValueType &trg = result->writable();
+		const typename From::ValueType &src = data->readable();
+
+		trg.reserve( m_mesh->numFaces() );
+
+		std::vector<int>::const_iterator vId = m_mesh->vertexIds()->readable().begin();
```

Comment:
I'll add a TODO for now and address in a follow up PR if @andrewkaufman  agrees?